cpp/src/column/column_view.cpp

```diff
@@ -76,6 +77,24 @@ size_type column_view_base::null_count(size_type begin, size_type end) const
            ? 0
            : cudf::count_unset_bits(null_mask(), offset() + begin, offset() + end);
 }
+
+// simple prime number multiplication algorithm.
+// Adapted from http://myeyesareblind.com/2017/02/06/Combine-hash-values/#apachecommons
+constexpr void combine_hash(size_t& h1, size_t h2) { h1 = h1 * 127 + h2; }
+
+size_t shallow_hash(column_view const& input)
+{
+  size_t hash = 0;
+  combine_hash(hash, std::hash<data_type>{}(input.type()));
+  combine_hash(hash, std::hash<size_type>{}(input.size()));
+  combine_hash(hash, std::hash<void const*>{}(input.head()));
+  combine_hash(hash, std::hash<void const*>{}(input.null_mask()));
```

Comment:
I don't think you should include the `head` or the `null_mask` pointer in the hash when the size is 0. Those pointers could be garbage if the column doesn't have any elements. In fact, you should explicitly add a test for this.

Reply:
then `slice(col_1, {0,0})[0] == slice(col_2, {0,0})[0]`. Is this correct behavior?

Reply:
I'm not sure I understand your example. Are you saying two columns without any elements will produce the same hash value? I think that's okay. With an empty column, there is no physical column that we're viewing, only a conceptual one. So all empty column_views of the same type conceptually view the same column.

Reply:
Yes. That's what I meant.
`shallow_hash(slice(column_view(col_1), {0,0})[0]) == shallow_hash(slice(column_view(col_1_copy), {0,0})[0])` fails for nested types only because children sizes may not be zero after slicing even if parent size is zero, and hence child.data is compared and they are different.

```
    auto col_new        = std::make_unique<cudf::column>(*col);
    auto col_new_view   = col_new->view();
    auto col_sliced     = cudf::slice(col_view, {0, 0, 1, 1, col_view.size(), col_view.size()});
    auto col_new_sliced = cudf::slice(col_new_view, {0, 0, 1, 1, col_view.size(), col_view.size()});
    EXPECT_EQ(shallow_hash(col_sliced[0]), shallow_hash(col_new_sliced[0]));
```

`parent.size()` could be propagated to  `shallow_hash(children)` too, but it does not seem right. (propagate one level or all the way to leaf). 
We can't simply ignore children altogether because 2 struct column view with different children types should be assumed to have different hash.

Reply:
I'm not suggesting you ignore children, but instead, ignore the values of `data()` and `null_mask()` when the column's size is zero. If two columns `c0, c1` both have `c0.size() == 0 == c1.size()` but they have children with different sizes/types/offsets/etc. then they are not shallow equal.

Reply:
commenting outcome of offline discussion: 
* for empty column, ignore the children, check types are same, ignore size, offset, pointers.
only look at the nested data types.
* call it `is_shallow_equivalent` instead of `is_shallow_equal`.